In [611]:
# Titanic Datasets
# Python imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [612]:
# reading the datasets
train_df = pd.read_csv('train.csv')
test_df = pd.read_csv('test.csv')

In [613]:
train_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [614]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


The training datasets contain 11 features and 1 survived class. Also it contain total 891 value including the some null values.In addition, there are 2 float values, 5 integer value and 5 categorical objects in the training datasets.

In [615]:
train_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [616]:
# let check the null values
# 38% survived in titanic.
total = train_df.isnull().sum().sort_values(ascending=False)
print total.head()

Cabin       687
Age         177
Embarked      2
Fare          0
Ticket        0
dtype: int64


In [617]:
len(train_df)

891

In [618]:
percent1 = train_df.isnull().sum()/len(train_df) * 100

In [619]:
percent1.head()

PassengerId    0.0
Survived       0.0
Pclass         0.0
Name           0.0
Sex            0.0
dtype: float64

In [620]:
missing_data = pd.concat([total,percent1], axis=1,keys=['Total','%'], sort=False)

In [621]:
missing_data.head()

Total          %
Cabin       687  77.104377
Age         177  19.865320
Embarked      2   0.224467
Fare          0   0.000000
Ticket        0   0.000000

The Embarked feature had only 2 missing values in comparison with Age 177 values and Cabin 687 values.
'Cabin' feature have 77% missing data.

In [622]:
# column values
train_df.columns.values

array(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'], dtype=object)

The above all feature will contribute to the survival features. But the PassengerId and Name and Ticket will not contribute more in the training. As a result, we will drop these features.

In [623]:
# drop the passengerID
train_df.drop(['PassengerId'], axis=1,inplace=True)
test_df.drop(['PassengerId'], axis=1)

Pclass                                               Name     Sex   Age  \
0         3                                   Kelly, Mr. James    male  34.5   
1         3                   Wilkes, Mrs. James (Ellen Needs)  female  47.0   
2         2                          Myles, Mr. Thomas Francis    male  62.0   
3         3                                   Wirz, Mr. Albert    male  27.0   
4         3       Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0   
5         3                         Svensson, Mr. Johan Cervin    male  14.0   
6         3                               Connolly, Miss. Kate  female  30.0   
7         2                       Caldwell, Mr. Albert Francis    male  26.0   
8         3          Abrahim, Mrs. Joseph (Sophie Halaut Easu)  female  18.0   
9         3                            Davies, Mr. John Samuel    male  21.0   
10        3                                   Ilieff, Mr. Ylio    male   NaN   
11        1                         Jones, Mr. Charles Cresson    male  46.0   
12        1      Snyder, Mrs. John Pillsbury (Nelle Stevenson)  female  23.0   
13        2                               Howard, Mr. Benjamin    male  63.0   
14        1  Chaffee, Mrs. Herbert Fuller (Carrie Constance...  female  47.0   
15        2      del Carlo, Mrs. Sebastiano (Argenia Genovesi)  female  24.0   
16        2                                  Keane, Mr. Daniel    male  35.0   
17        3                                  Assaf, Mr. Gerios    male  21.0   
18        3                       Ilmakangas, Miss. Ida Livija  female  27.0   
19        3              Assaf Khalil, Mrs. Mariana (Miriam")"  female  45.0   
20        1                             Rothschild, Mr. Martin    male  55.0   
21        3                          Olsen, Master. Artur Karl    male   9.0   
22        1               Flegenheim, Mrs. Alfred (Antoinette)  female   NaN   
23        1                    Williams, Mr. Richard Norris II    male  21.0   
24        1    Ryerson, Mrs. Arthur Larned (Emily Maria Borie)  female  48.0   
25        3                            Robins, Mr. Alexander A    male  50.0   
26        1                       Ostby, Miss. Helene Ragnhild  female  22.0   
27        3                                  Daher, Mr. Shedid    male  22.5   
28        1                            Brady, Mr. John Bertram    male  41.0   
29        3                                  Samaan, Mr. Elias    male   NaN   
..      ...                                                ...     ...   ...   
388       3                               Canavan, Mr. Patrick    male  21.0   
389       3                        Palsson, Master. Paul Folke    male   6.0   
390       1                         Payne, Mr. Vivian Ponsonby    male  23.0   
391       1     Lines, Mrs. Ernest H (Elizabeth Lindsey James)  female  51.0   
392       3                      Abbott, Master. Eugene Joseph    male  13.0   
393       2                               Gilbert, Mr. William    male  47.0   
394       3                           Kink-Heilmann, Mr. Anton    male  29.0   
395       1     Smith, Mrs. Lucien Philip (Mary Eloise Hughes)  female  18.0   
396       3                               Colbert, Mr. Patrick    male  24.0   
397       1  Frolicher-Stehli, Mrs. Maxmillian (Margaretha ...  female  48.0   
398       3                     Larsson-Rondberg, Mr. Edvard A    male  22.0   
399       3                           Conlon, Mr. Thomas Henry    male  31.0   
400       1                            Bonnell, Miss. Caroline  female  30.0   
401       2                                    Gale, Mr. Harry    male  38.0   
402       1                     Gibson, Miss. Dorothy Winifred  female  22.0   
403       1                             Carrau, Mr. Jose Pedro    male  17.0   
404       1                       Frauenthal, Mr. Isaac Gerald    male  43.0   
405       2       Nourney, Mr. Alfred (Baron von Drachstedt")"    male  20.0   
406  

In [624]:
train_df.columns.values

array(['Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'], dtype=object)

In [625]:
train_df[["Cabin","Survived"]].head()

Cabin  Survived
0   NaN         0
1   C85         1
2   NaN         1
3  C123         1
4   NaN         0

In [626]:
# import re
# deck = {"A": 1, "B": 2, "C": 3, "D": 4, "E": 5, "F": 6, "G": 7, "U": 8}
# data = [train_df, test_df]

# for dataset in data:
#     dataset['Cabin'] = dataset['Cabin'].fillna("U0")
#     dataset['Deck'] = dataset['Cabin'].map(lambda x: re.compile("([a-zA-Z]+)").search(x).group())
#     dataset['Deck'] = dataset['Deck'].map(deck)
#     dataset['Deck'] = dataset['Deck'].fillna(0)
#     dataset['Deck'] = dataset['Deck'].astype(int) 

In [627]:
import re
deck = {'A':1,'B':2,'C':3,'D':4,'E':5,'F':6,'G':7,'U':8}
data = [train_df,test_df]
for value  in data:
    value["Cabin"] = value["Cabin"].fillna('U0')
    value["Deck"] = value["Cabin"].map(lambda x: re.compile("([a-zA-Z]+)").search(x).group())
    value["Deck"] = value["Deck"].map(deck)
    value["Deck"] = value["Deck"].fillna(0)
    value["Deck"] = value["Deck"].astype('int')
    #print value
    #value["Cabin"]

In [628]:
value.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  Deck  
0  34.5      0      0   330911   7.8292    U0        Q     8  
1  47.0      1      0   363272   7.0000    U0        S     8  
2  62.0      0      0   240276   9.6875    U0        Q     8  
3  27.0      0      0   315154   8.6625    U0        S     8  
4  22.0      1      1  3101298  12.2875    U0        S     8

In [629]:
# check the null values
total = train_df.isnull().sum().sort_values(ascending=False)


In [630]:
total

Age         177
Embarked      2
Deck          0
Cabin         0
Fare          0
Ticket        0
Parch         0
SibSp         0
Sex           0
Name          0
Pclass        0
Survived      0
dtype: int64

Cabin have 0 null values. Let edit Embarked features

In [631]:
# drop the Cabin values
train_df.drop(['Cabin'], axis=1,inplace=True)
test_df.drop(['Cabin'], axis=1,inplace=True)

In [632]:
train_df.columns.values

array(['Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Embarked', 'Deck'], dtype=object)

Let remove the AGE features null values

In [633]:
# data = [train_df, test_df]

# for dataset in data:
#     mean = train_df["Age"].mean()
#     std = test_df["Age"].std()
#     is_null = dataset["Age"].isnull().sum()
#     # compute random numbers between the mean, std and is_null
#     rand_age = np.random.randint(mean - std, mean + std, size = is_null)
#     # fill NaN values in Age column with random values generated
#     age_slice = dataset["Age"].copy()
#     age_slice[np.isnan(age_slice)] = rand_age
#     dataset["Age"] = age_slice
#     dataset["Age"] = train_df["Age"].astype(int)

In [634]:
data = [train_df, test_df]
for value in data:
    mean = value["Age"].mean()
    std = value["Age"].std()
    is_null = value["Age"].isnull().sum()
    print 'Mean:-- {}, STD:-- {}, NULL:-- {}'.format(mean,std,is_null)
    # compute the random values
    rand_age = np.random.randint(mean-std,mean + std,size=is_null)
    # print rand_age
    # fill the null values
    age_slice = value["Age"].copy()
    #print age_slice
    age_slice[np.isnan(age_slice)] = rand_age
    #print age_slice
    value["Age"] = age_slice
    value["Age"] = value["Age"].astype(int)

Mean:-- 29.6991176471, STD:-- 14.5264973323, NULL:-- 177
Mean:-- 30.2725903614, STD:-- 14.1812092356, NULL:-- 86


In [635]:
train_df.head()

Survived  Pclass                                               Name  \
0         0       3                            Braund, Mr. Owen Harris   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   

      Sex  Age  SibSp  Parch            Ticket     Fare Embarked  Deck  
0    male   22      1      0         A/5 21171   7.2500        S     8  
1  female   38      1      0          PC 17599  71.2833        C     3  
2  female   26      0      0  STON/O2. 3101282   7.9250        S     8  
3  female   35      1      0            113803  53.1000        S     3  
4    male   35      0      0            373450   8.0500        S     8

In [636]:
# Now the number of null values are:--
train_df.isnull().sum().sort_values(ascending=False)


Embarked    2
Deck        0
Fare        0
Ticket      0
Parch       0
SibSp       0
Age         0
Sex         0
Name        0
Pclass      0
Survived    0
dtype: int64

Only Embarked feature have null values in the datasets


In [637]:
train_df["Embarked"].head()

0    S
1    C
2    S
3    S
4    S
Name: Embarked, dtype: object

In [638]:
train_df["Embarked"].describe()

count     889
unique      3
top         S
freq      644
Name: Embarked, dtype: object

In [639]:
# replace the null value with the common value 'S'
data = [train_df, test_df]
for value in data:
    value["Embarked"] = value["Embarked"].fillna('S')

In [640]:
train_df.isnull().sum().sort_values(ascending=False)


Deck        0
Embarked    0
Fare        0
Ticket      0
Parch       0
SibSp       0
Age         0
Sex         0
Name        0
Pclass      0
Survived    0
dtype: int64

Hence all the null values are converted into non-null values.

In [641]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
Survived    891 non-null int64
Pclass      891 non-null int64
Name        891 non-null object
Sex         891 non-null object
Age         891 non-null int64
SibSp       891 non-null int64
Parch       891 non-null int64
Ticket      891 non-null object
Fare        891 non-null float64
Embarked    891 non-null object
Deck        891 non-null int64
dtypes: float64(1), int64(6), object(4)
memory usage: 76.6+ KB


Still there are float, int and object features.

In [642]:
train_df["Fare"].head(10)

0     7.2500
1    71.2833
2     7.9250
3    53.1000
4     8.0500
5     8.4583
6    51.8625
7    21.0750
8    11.1333
9    30.0708
Name: Fare, dtype: float64

In [643]:
data = [train_df, test_df]
for value in data:
    value['Fare'] = value['Fare'].fillna(value["Fare"].median())
    value['Fare'] = value['Fare'].astype(int)

Remaining are the categorical feature:- Name, Ticket, Embarked, and Sex

Changing the Sex Feature

In [644]:
train_df["Sex"].head()


0      male
1    female
2    female
3    female
4      male
Name: Sex, dtype: object

In [645]:
gender = {'male':0, 'female':1}
data = [train_df, test_df]
for value in data:
    value["Sex"] = value["Sex"].map(gender)


In [646]:
train_df.head(10)
             

Survived  Pclass                                               Name  Sex  \
0         0       3                            Braund, Mr. Owen Harris    0   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...    1   
2         1       3                             Heikkinen, Miss. Laina    1   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)    1   
4         0       3                           Allen, Mr. William Henry    0   
5         0       3                                   Moran, Mr. James    0   
6         0       1                            McCarthy, Mr. Timothy J    0   
7         0       3                     Palsson, Master. Gosta Leonard    0   
8         1       3  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)    1   
9         1       2                Nasser, Mrs. Nicholas (Adele Achem)    1   

   Age  SibSp  Parch            Ticket  Fare Embarked  Deck  
0   22      1      0         A/5 21171     7        S     8  
1   38      1      0          PC 17599    71        C     3  
2   26      0      0  STON/O2. 3101282     7        S     8  
3   35      1      0            113803    53        S     3  
4   35      0      0            373450     8        S     8  
5   27      0      0            330877     8        Q     8  
6   54      0      0             17463    51        S     5  
7    2      3      1            349909    21        S     8  
8   27      0      2            347742    11        S     8  
9   14      1      0            237736    30        C     8

Changing the Name Feature

In [647]:
# dropping the Name features
train_df.drop(['Name'],inplace=True, axis=1)
test_df.drop(['Name'],inplace=True, axis=1)

In [648]:
train_df.columns.values


array(['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket',
       'Fare', 'Embarked', 'Deck'], dtype=object)

In [649]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
Survived    891 non-null int64
Pclass      891 non-null int64
Sex         891 non-null int64
Age         891 non-null int64
SibSp       891 non-null int64
Parch       891 non-null int64
Ticket      891 non-null object
Fare        891 non-null int64
Embarked    891 non-null object
Deck        891 non-null int64
dtypes: int64(8), object(2)
memory usage: 69.7+ KB


In [650]:
train_df["Embarked"].head()

0    S
1    C
2    S
3    S
4    S
Name: Embarked, dtype: object

In [651]:
values = {'S':0,'C':1,'Q':2}
data = [train_df,test_df]
for value in data:
    value["Embarked"] = value["Embarked"].map(values)

In [652]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
Survived    891 non-null int64
Pclass      891 non-null int64
Sex         891 non-null int64
Age         891 non-null int64
SibSp       891 non-null int64
Parch       891 non-null int64
Ticket      891 non-null object
Fare        891 non-null int64
Embarked    891 non-null int64
Deck        891 non-null int64
dtypes: int64(9), object(1)
memory usage: 69.7+ KB


In [653]:
train_df["Ticket"].describe()

count          891
unique         681
top       CA. 2343
freq             7
Name: Ticket, dtype: object

IT has 682 unique values, So we will drop the Ticket features

In [654]:
train_df.drop(['Ticket'],axis=1,inplace=True)
test_df.drop(['Ticket'],axis=1,inplace=True)

In [655]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
Survived    891 non-null int64
Pclass      891 non-null int64
Sex         891 non-null int64
Age         891 non-null int64
SibSp       891 non-null int64
Parch       891 non-null int64
Fare        891 non-null int64
Embarked    891 non-null int64
Deck        891 non-null int64
dtypes: int64(9)
memory usage: 62.7 KB


In [656]:
train_df.head()

Survived  Pclass  Sex  Age  SibSp  Parch  Fare  Embarked  Deck
0         0       3    0   22      1      0     7         0     8
1         1       1    1   38      1      0    71         1     3
2         1       3    1   26      0      0     7         0     8
3         1       1    1   35      1      0    53         0     3
4         0       3    0   35      0      0     8         0     8

In [657]:
# Training the data into models
Y_train = train_df["Survived"]
X_train = train_df.drop(["Survived"], axis=1)
X_test.head()

Pclass  Sex  Age  SibSp  Parch  Fare  Embarked  Deck
0       3    0   34      0      0     7         2     8
1       3    1   47      1      0     7         0     8
2       2    0   62      0      0     9         2     8
3       3    0   27      0      0     8         0     8
4       3    1   22      1      1    12         0     8

In [658]:
test_df.columns.values

array(['PassengerId', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare',
       'Embarked', 'Deck'], dtype=object)

In [659]:
# Python imports from sklearn
from sklearn import linear_model
from sklearn.svm import SVC, LinearSVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

In [660]:
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import  GaussianNB
from sklearn.linear_model import Perceptron
from sklearn.linear_model import SGDClassifier

In [661]:
# Random Forest
regressor = RandomForestClassifier(n_estimators=30, random_state = 0)
regressor.fit(X_train, Y_train)
X_test = X_test.drop(["Ticket"], axis=1).copy()
X_test.head()

KeyError: "['Ticket'] not found in axis"

In [ ]:
X_train.head()

In [ ]:
y_pred = regressor.predict(X_test)
acc_random_forest = round(regressor.score(X_train, Y_train)*100,2)
random_score



In [ ]:
# Logistic Regression
log_regressor = LogisticRegression()
log_regressor.fit(X_train, Y_train)
y_pred = log_regressor.predict(X_test)

In [ ]:
acc_log_regression = round(log_regressor.score(X_train, Y_train)*100,2)

In [ ]:
# K neighbour Neihgbours
from sklearn.neighbors import KNeighborsClassifier


In [ ]:
knn = KNeighborsClassifier(n_neighbors=3)
knn.fit(X_train, Y_train)
y_pred = knn.predict(X_test)

In [ ]:
acc_knn = round(knn.score(X_train, Y_train)*100,2)

In [ ]:
#gaussian Naiove bayes
gaussian =GaussianNB()
gaussian.fit(X_train, Y_train)
y_pred = gaussian.predict(X_test)

In [ ]:
acc_gaussian = round(gaussian.score(X_train, Y_train)*100,2)

In [ ]:
#decisionTree
dtree = DecisionTreeClassifier()
dtree.fit(X_train, Y_train)
dtree.predict(X_test)
acc_dtree = round(dtree.score(X_train, Y_train) * 100,2)

DECIDE THE BEST MODEL

In [ ]:
final_score = [acc_dtree, acc_gaussian, acc_knn, acc_log_regression, acc_random_forest]

In [ ]:
final_score

In [ ]:
model_name = ['DecisionTree', 'Gaussian','Kneighbors','LogisticRegression','RandomForest']

In [ ]:
model_name

In [ ]:
pd.DataFrame({'Model_Name':model_name,'FInal_score':final_score})

Random Forest and Decision Tree have higher accuracy

In [ ]:
# confusion matrix
random_forest = RandomForestClassifier(n_estimators=100, random_state =1)
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import confusion_matrix
predictions = cross_val_predict(random_forest, X_train, Y_train, cv=3)
#print predictions


In [ ]:
confusion_matrix(Y_train,predictions)

In [ ]:
random_forest.fit(X_train, Y_train)
y_predictions = random_forest.predict(X_test)

In [ ]:
random_forest.score(X_train, Y_train)


In [664]:
submission = pd.DataFrame({
        "PassengerId": test_df["PassengerId"],
        "Survived": y_predictions
    })

In [665]:
submission.head()

PassengerId  Survived
0          892         0
1          893         0
2          894         0
3          895         1
4          896         0